Firestore/Source/Local/FSTLocalStore.mm

```diff
@@ -156,12 +159,51 @@ - (FSTMaybeDocumentDictionary *)userDidChange:(const User &)user {
 }
 
 - (FSTLocalWriteResult *)locallyWriteMutations:(NSArray<FSTMutation *> *)mutations {
+  FIRTimestamp *localWriteTime = [FIRTimestamp timestamp];
+  DocumentKeySet keys;
+  for (FSTMutation *mutation in mutations) {
+    keys = keys.insert(mutation.key);
+  }
+
   return self.persistence.run("Locally write mutations", [&]() -> FSTLocalWriteResult * {
-    FIRTimestamp *localWriteTime = [FIRTimestamp timestamp];
-    FSTMutationBatch *batch =
-        [self.mutationQueue addMutationBatchWithWriteTime:localWriteTime mutations:mutations];
-    DocumentKeySet keys = [batch keys];
-    FSTMaybeDocumentDictionary *changedDocuments = [self.localDocuments documentsForKeys:keys];
+    // Load and apply all existing mutations. This lets us compute the current base state for
+    // all non-idempotent transforms before applying any additional user-provided writes.
+    FSTMaybeDocumentDictionary *existingDocuments = [self.localDocuments documentsForKeys:keys];
+
+    // For non-idempotent mutations (such as `FieldValue.increment()`), we record the base
+    // state in a separate patch mutation. This is later used to guarantee consistent values
+    // and prevents flicker even if the backend sends us an update that already includes our
+    // transform.
+    NSMutableArray<FSTMutation *> *baseMutations = [NSMutableArray array];
+    for (FSTMutation *mutation in mutations) {
+      FSTMaybeDocument *maybeDocument = [existingDocuments objectForKey:mutation.key];
+      if (!mutation.idempotent) {
+        // Theoretically, we should only include non-idempotent fields in this field mask as
```

Comment:
Sorry, I'm not sure I understand the comment. The way I'm reading it, it seems to say that ideally, we would only include non-idempotent fields, which is sufficient to prevent flicker. However, we also include idempotent fields, which has a downside. What I don't understand is why we include all fields if it has a downside and including only non-idempotent fields is sufficient?

Reply:
Your understanding is 100% correct. The reason behind this comment is that I prefer the simplicity of having a `getFieldMask()` method over one that filters by idempotency. I added a TODO to maybe re-evaluate this in the future.